### PR TITLE
Add ability to use `TypedDict` `Required/NotRequired`

### PR DESCRIPTION
**Pull request by uriyyo**
## Add ability to use TypedDict Required/NotRequired

Recently `Required` and `NotRequire` was added to `typing_extentions`.
This PR adds ability to use them in `TypedDict` declaration:
```py
from typing_extensions import NotRequired, Required, TypedDict
from pydantic import BaseModel


class DataTD1(TypedDict, total=False):
    a: int
    b: Required[str]


class DataTD2(TypedDict, total=True):
    a: NotRequired[int]
    b: str


class Model1(BaseModel):
    t1: DataTD1
    t2: DataTD2


Model1(t1={}, t2={})
```
```py
pydantic.error_wrappers.ValidationError: 2 validation errors for Model1
t1 -> b
  field required (type=value_error.missing)
t2 -> b
  field required (type=value_error.missing)
```

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [ ] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by uriyyo** [comment]
please review

File: tests/test_annotated_types.py [hunk 1]
```diff
@@ -290,3 +290,54 @@ class Model(BaseModel):
 
     with pytest.raises(ValidationError):
         Model.parse_obj({'t': {'v': -1}})
+
+
+def test_typeddict_required_not_required():
```
Comment (samuelcolvin) [inline tests/test_annotated_types.py:295]: can we separate this into multiple tests.

File: pydantic/typing.py [hunk 2]
```diff
@@ -363,6 +368,38 @@ def is_typeddict(type_: Type[Any]) -> bool:
     return lenient_issubclass(type_, dict) and hasattr(type_, '__total__')
 
 
+def is_typeddict_special(type_: Type[Any]) -> bool:
+    """
+    Check if type is a TypedDict special form (Required or NotRequired).
+    """
+    return type_ is TypedDictRequired or type_ is TypedDictNotRequired
```
Comment (samuelcolvin) [inline pydantic/typing.py:375]:
```suggestion
def is_typeddict_special(type_: Any) -> bool:
    """
    Check if type is a TypedDict special form (Required or NotRequired).
    """
    return type_ is not None and (type_ is TypedDictRequired or type_ is TypedDictNotRequired)
```

maybe? would make the usage above slightly cleaner.

File: pydantic/typing.py [hunk 3]
```diff
@@ -363,6 +368,38 @@ def is_typeddict(type_: Type[Any]) -> bool:
...
+
+
+def update_typeddict_keys(
+    typeddict_cls: Type['TypedDict'],  # type: ignore[valid-type]
+) -> None:
```
Comment (samuelcolvin) [inline pydantic/typing.py:380]: how come this is required if `typeddict_cls.__required_keys__` and `typeddict_cls.__optional_keys__` are already set?

Might be worth adding an explanation to the docstring.

**Review by samuelcolvin — COMMENTED** [review]
overall this looks great. @PrettyWood since you did the initial `TypedDict` stuff, what do you think?

**Comment by samuelcolvin** [comment]
please update.

File: tests/test_annotated_types.py (hunk 1 above)
Comment (uriyyo) [reply tests/test_annotated_types.py]: fixed, thanks

File: pydantic/typing.py (hunk 3 above)
Comment (uriyyo) [reply pydantic/typing.py:380]: Please, review new doc string.

**Comment by uriyyo** [comment]
please review

**Comment by uriyyo** [comment]
please review

File: pydantic/typing.py [hunk 4]
```diff
@@ -435,6 +440,42 @@ def is_typeddict(type_: Type[Any]) -> bool:
     return lenient_issubclass(type_, dict) and hasattr(type_, '__total__')
 
 
+def is_typeddict_special(type_: Any) -> bool:
+    """
+    Check if type is a TypedDict special form (Required or NotRequired).
+    """
+    return type_ is not None and (type_ is TypedDictRequired or type_ is TypedDictNotRequired)
```
Comment (PrettyWood) [inline pydantic/typing.py:447]:
```suggestion
    return type_ is TypedDictRequired or type_ is TypedDictNotRequired
```
is not enough?

File: pydantic/typing.py (hunk 3 above)
Comment (PrettyWood) [reply pydantic/typing.py:452]: I'm not really for it. The issue comes from `typing.TypedDict` in 3.10, which doesn't support that.
But `TypedDict` from `typing` in 3.11 and `typing_extensions` supports it.
Couldn't we just warn the user to not use `typing.TypedDict` if the version is `3.10` and `Required` or `NotRequired` is used?

**Comment by uriyyo** [comment]
please review

**Pull request merged**